demos/qml_optimization.ipynb

```diff
@@ -0,0 +1,469 @@
+      "execution_count": 7,
+      "metadata": {
+        "id": "sOWgLW5esIsO"
+      },
+      "outputs": [],
+      "source": [
+        "weights = jnp.ones([n_wires, 3])\n",
+        "bias = jnp.array(0.)\n",
+        "params = {\"weights\": weights, \"bias\": bias}"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "4mREqzeb7LI7"
+      },
+      "source": [
+        "Plugging the trainable parameters, data, and target labels into our cost function, we can see the current loss as well as the parameter gradients:"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 8,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "23t_bcyOtNBV",
+        "outputId": "7022b2d4-6c12-4a25-f903-00fd4e51be04"
+      },
+      "outputs": [
+        {
+          "data": {
+            "text/plain": [
+              "array(0.29232612)"
+            ]
+          },
+          "execution_count": 8,
+          "metadata": {},
+          "output_type": "execute_result"
+        }
+      ],
+      "source": [
+        "loss_fn(params, data, targets)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 9,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "XCeL2_c4tPSi",
+        "outputId": "f405b448-494c-4e67-f187-e7a3adfdfe7b"
+      },
+      "outputs": [
+        {
+          "name": "stdout",
+          "output_type": "stream",
+          "text": [
+            "{'bias': array(-0.75432067), 'weights': array([[-1.95077271e-01,  5.28546590e-02, -4.89252073e-01],\n",
+            "       [-1.99687794e-02, -5.32871564e-02,  9.22904864e-02],\n",
+            "       [-2.71755507e-03, -9.64672786e-05, -4.79570827e-03],\n",
+            "       [-6.35443870e-02,  3.61110014e-02, -2.05196876e-01],\n",
+            "       [-9.02635405e-02,  1.63759364e-01, -5.64262612e-01]])}\n"
+          ]
+        }
+      ],
+      "source": [
+        "print(qml.qjit(catalyst.grad(loss_fn, method=\"fd\"))(params, data, targets))"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "yzRbmfJv7PC_"
+      },
+      "source": [
+        "## Create the optimizer\n",
+        "\n",
+        "We can now use Optax to create an Adam optimizer, and train our circuit.\n",
+        "\n",
+        "We first define our `update_step` function, which needs to do a couple of things:\n",
+        "\n",
+        "- Compute the loss function (so we can track training) and the gradients (so we can apply an optimization step). We can do this in one execution via the `jax.value_and_grad` function.\n",
+        "\n",
+        "- Apply the update step of our optimizer via `opt.update`\n",
+        "\n",
+        "- Update the parameters via `optax.apply_updates`\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 35,
```

Comment:
Would you re-run this notebook to get all these execution counts in order?

Reply:
yep can do